lectures/week5/reader.cpp

```diff
@@ -0,0 +1,79 @@
+#include <iostream>
+#include <memory>
+#include <vector>
+#include <sstream>
+#include <cstdio>
+#include <algorithm>
+#include <ostream>
+#include <iterator>
+#include <string>
+
+class Reader {
+ public:
+  Reader(std::shared_ptr<std::string> text): text_(text), location_{0} {}
+  void advance() {
+    ++location_;
+  }
+  bool atEnd() const {
+    return location_ == text_->size();
+  }
+  std::weak_ptr<std::string> getText() const {
+    return text_;

```

Comment:
I'm strongly against this. This isn't a useful use case for weak pointers. This would be far better off with unique pointers + raw pointers.
Weak pointers are only valuable if  you're not sure if the object is still alive. This is not the case here.